components/src/motion/pmw3901/pmw3901.adb

```diff
@@ -0,0 +1,310 @@
+                     (16#7f#, 16#09#),
+                     (16#4f#, 16#af#),
+                     (16#5f#, 16#40#),
+                     (16#48#, 16#80#),
+                     (16#49#, 16#80#),
+                     (16#57#, 16#77#),
+                     (16#60#, 16#78#),
+                     (16#61#, 16#78#),
+                     (16#62#, 16#08#),
+                     (16#63#, 16#50#),
+                     (16#7f#, 16#0a#),
+                     (16#45#, 16#60#),
+                     (16#7f#, 16#00#),
+                     (16#4d#, 16#11#),
+                     (16#55#, 16#80#),
+                     (16#74#, 16#21#),
+                     (16#75#, 16#1f#),
+                     (16#4a#, 16#78#),
+                     (16#4b#, 16#78#),
+                     (16#44#, 16#08#),
+                     (16#45#, 16#50#),
+                     (16#64#, 16#ff#),
+                     (16#65#, 16#1f#),
+                     (16#7f#, 16#14#),
+                     (16#65#, 16#67#),
+                     (16#66#, 16#08#),
+                     (16#63#, 16#70#),
+                     (16#7f#, 16#15#),
+                     (16#48#, 16#48#),
+                     (16#7f#, 16#07#),
+                     (16#41#, 16#0d#),
+                     (16#43#, 16#14#),
+                     (16#4b#, 16#0e#),
+                     (16#45#, 16#0f#),
+                     (16#44#, 16#42#),
+                     (16#4c#, 16#80#),
+                     (16#7f#, 16#10#),
+                     (16#5b#, 16#02#),
+                     (16#7f#, 16#07#),
+                     (16#40#, 16#41#),
+                     (16#70#, 16#00#)));
+      This.Timing.Delay_Milliseconds (16#0a#);
+      Write_Magics (((16#32#, 16#44#),
+                     (16#7f#, 16#07#),
+                     (16#40#, 16#40#),
+                     (16#7f#, 16#06#),
+                     (16#62#, 16#f0#),
+                     (16#63#, 16#00#),
+                     (16#7f#, 16#0d#),
+                     (16#48#, 16#c0#),
+                     (16#6f#, 16#d5#),
+                     (16#7f#, 16#00#),
+                     (16#5b#, 16#a0#),
+                     (16#4e#, 16#a8#),
+                     (16#5a#, 16#50#),
+                     (16#40#, 16#80#)));
+      This.Timing.Delay_Milliseconds (16#f0#);
+      --  enable LED pulsing???
+      Write_Magics (((16#7f#, 16#14#),
+                     (16#6f#, 16#1c#),
+                     (16#7f#, 16#00#)));
+   end Calibrate;
+
+   function Read_Motion  (This : in out PMW3901_Flow_Sensor) return Motion is
+      subtype Buffer is STM32.SPI.UInt8_Buffer (1 .. 12);
+      function Convert is new Ada.Unchecked_Conversion (Buffer, Motion);
+      Dummy : HAL.UInt8;
+      Buff : Buffer := (others => 0);
+   begin
+      This.CS.Drive (False);
+      This.Port.Transmit_Receive (Outgoing => Reg_Motion_Burst,
+                                 Incoming => Dummy);
+      pragma Warnings (Off, "referenced before it has a value");
```

Comment:
This is a pretty important warning you are silencing. Maybe you can have a different buffer incoming and outgoing.

Reply:
For some reason, all the code I’ve seen insisted on using this form (transmit  request, receive dummy, transmit dummy, receive result of request). Why, I don’t know, because doing it as you would have expected (transmit the request, receive the result) works just fine; new code doesn’t provoke this warning.